052-record-reconciled-versions.md

```diff
@@ -0,0 +1,190 @@
+
+This proposal assumes that all key value pairs mentioned will be writen to the `status` field as this is managed by the operator and can be added to without impacting end users.
+
+The following are the proposed new fields:
+
+### status.operatorLastSuccessfulVersion
+
+The `status.operatorLastSuccessfulVersion` field is patched at the end of a **successful** reconcile, and signals to a user that the operator at version `X` was able to reconcile to successful completion, meaning it reached the end of a reconcile without error.
+This would be the field a user would query against to check whether an operator upgrade was complete - or at least a complete roll of the Kafka brokers on the latest operator version. Details on how to verify that an upgrade has completed when the `Kafka.spec.kafka.version` is updated is covered in the next section.
+
+Example field in status:
+```
+status:
+  operatorLastSuccessfulVersion: '0.37.0'
+```
+If a reconcile does not reach completion, this field is not added or updated, however in the case where it already exists it is left at the previous value.
+i.e.
+`status.operatorLastSuccessfulVersion=0.35.0` would remain if a post-operator-upgrade reconciliation on `0.36.0` failed, as the information that a prior reconciliation passed would be useful to an end user.
+
+### status.kafkaVersion
+
+The `status.kafkaVersion` field is updated in the CR at the end of the KafkaReconciler in the `podsReady` method and signals to a user that the KafkaReconciler has finished reconciling the Kafka brokers to Kafka version `Y`, regardless of success or failure later in the reconcile.
+```
+status:
+  kafkaVersion '3.5.1'
+```
+
+Note: This will update the field in the reconcilers internal status object and have it be written at the end of the reconciliation rather then having a separate status update API call each time this is updated.
+
+This field is required as it signals that a reconcile to Kafka version `Y` has finished, and this extra information is important and needs to be available, as a user just checking `status.operatorLastSuccessfulVersion` is not sufficient for upgrades where multiple rolls of the brokers are required, such as a Kafka version update alongside an upgrade.
+
+If the operator errors during the Kafka reconciliation this field will not be updated, and will instead remain set to its prior set value, or unset if no previous value was set.
+This does mean that during an upgrade the value set in the `status.kafkaVersion` field will not reflect the kafka version correctly for all brokers, but this is desired, as the operator should only update the status field to reflect a finished upgrade so that a user can query the field to understand when an upgrade is finished and not for a breakdown broker-by-broker, if a user wants an understanding this granular this proposal does not cover this, and would instead recommend they check the kafka broker versions by querying the pods.
```

Comment:
```suggestion
During an upgrade, the value in the `status.kafkaVersion` field may not accurately reflect the Kafka version for all brokers. However, this behavior is intentional. The operator updates the status field only upon completing an upgrade, allowing users to determine upgrade completion at the cluster level, rather than on a broker-by-broker basis. For a more granular check, users seeking individual broker versions can query the Kafka broker versions through pod queries.
```